crates/cdk/src/wallet/key_manager/mod.rs

```diff
@@ -0,0 +1,327 @@
+    /// and waits up to 5 seconds for the keys to arrive.
+    pub async fn get_keys(&self, keyset_id: &Id) -> Result<Arc<Keys>, Error> {
+        for _ in 0..MAX_RETRY {
+            let cache = self.cache.load();
+            if cache.is_ready {
+                return cache
+                    .keys_by_id
+                    .get(keyset_id)
+                    .cloned()
+                    .ok_or(Error::UnknownKeySet);
+            }
+            tokio::time::sleep(RETRY_SLEEP).await;
+        }
+
+        Err(Error::UnknownKeySet)
+    }
+
+    /// Get keyset info by ID (cache-first with automatic refresh)
+    pub async fn get_keyset_by_id(&self, keyset_id: &Id) -> Result<Arc<KeySetInfo>, Error> {
+        for _ in 0..MAX_RETRY {
+            let cache = self.cache.load();
+            if cache.is_ready {
+                return cache
+                    .keysets_by_id
+                    .get(keyset_id)
+                    .cloned()
+                    .ok_or(Error::UnknownKeySet);
+            }
+            tokio::time::sleep(RETRY_SLEEP).await;
+        }
+
+        Err(Error::UnknownKeySet)
+    }
+
+    /// Get all keysets for the mint (cache-first with automatic refresh)
+    pub async fn get_keysets(&self) -> Result<Vec<KeySetInfo>, Error> {
+        for _ in 0..MAX_RETRY {
+            let cache = self.cache.load();
+            if cache.is_ready {
+                let keysets: Vec<KeySetInfo> = cache
+                    .keysets_by_id
+                    .values()
+                    .map(|ks| (**ks).clone())
+                    .collect();
+                return if keysets.is_empty() {
+                    Err(Error::UnknownKeySet)
+                } else {
+                    Ok(keysets)
+                };
+            }
+
+            tokio::time::sleep(RETRY_SLEEP).await;
+        }
+
+        Err(Error::UnknownKeySet)
+    }
+
+    /// Get all active keysets for the mint (cache-first with automatic refresh)
+    pub async fn get_active_keysets(&self) -> Result<Vec<Arc<KeySetInfo>>, Error> {
+        for _ in 0..MAX_RETRY {
+            let cache = self.cache.load();
+            if cache.is_ready {
+                return Ok(cache.active_keysets.clone());
+            }
+            tokio::time::sleep(RETRY_SLEEP).await;
+        }
+
+        Err(Error::UnknownKeySet)
+    }
+
+    /// Trigger a refresh and wait for it to complete
+    ///
+    /// Sends a refresh message to the background task and waits up to 5 seconds
+    /// for the cache to be updated with a newer version.
+    pub async fn refresh(&self) -> Result<Vec<KeySetInfo>, Error> {
+        let last_version = self.cache.load().refresh_version;
+
+        self.send_message(MessageToWorker::FetchMint);
+
+        for _ in 0..MAX_RETRY {
+            if let Some(keysets) = {
+                let cache = self.cache.load();
+                if last_version < cache.refresh_version {
```

Comment:
This send_message will silently fail in the admittedly unlikely event the message cannot be sent on the channel but then we will still go into the polling loop. If we know the channel message failed we should return early with an error I think.